internal/ingress/metric/collector/stats.go

```diff
@@ -0,0 +1,181 @@
+	"net"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type udpData struct {
+	Host   string `json:"host"`   // Label
+	Status string `json:"status"` // Label
+
+	RemoteAddress string `json:"remoteAddr"` // Label
+	RemoteUser    string `json:"remoteUser"` // Label
+
+	BytesSent float64 `json:"bytesSent,string"` // Metric
+
+	Protocol string `json:"protocol"` // Label
+	Method   string `json:"method"`   // Label
+	Path     string `json:"path"`     // Label
+
+	RequestLength   float64 `json:"requestLength,string"`   // Metric
+	RequestDuration float64 `json:"requestDuration,string"` // Metric
+
+	UpstreamName         string  `json:"upstreamName"`                // Label
+	UpstreamIP           string  `json:"upstreamIP"`                  // Label
+	UpstreamResponseTime float64 `json:"upstreamResponseTime,string"` // Metric
+	UpstreamStatus       string  `json:"upstreamStatus"`              // Label
+
+	Namespace string `json:"namespace"` // Label
+	Ingress   string `json:"ingress"`   // Label
+	Service   string `json:"service"`   // Label
+}
+
+type StatsCollector struct {
+	upstreamResponseTime *prometheus.HistogramVec
+	requestDuration      *prometheus.HistogramVec
+	requestLength        *prometheus.HistogramVec
+	bytesSent            *prometheus.HistogramVec
+	listener             *net.UDPConn
+	ns                   string
+	watchClass           string
+	port                 int
+}
+
+func NewInstance(ns string, class string, port int) (*StatsCollector, error) {
+	sc := StatsCollector{}
+
+	listener, err := newUDPListener(port)
+
+	if err != nil {
+		return nil, err
+	}
+
+	sc.listener = listener
+	sc.ns = ns
+	sc.watchClass = class
+	sc.port = port
+
+	tags := []string{"host", "status", "remote_address", "remote_user", "protocol", "method", "path", "upstream_name", "upstream_ip", "upstream_status", "namespace", "ingress", "service"}
+
+	sc.upstreamResponseTime = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:      "upstream_response_time_seconds",
+			Help:      "The time spent on receiving the response from the upstream server",
+			Namespace: ns,
+		},
+		tags,
+	)
+
+	sc.requestDuration = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:      "request_duration_seconds",
+			Help:      "The request processing time in seconds",
+			Namespace: ns,
+		},
+		tags,
+	)
+
+	sc.requestLength = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:      "request_length_bytes",
+			Help:      "The request length (including request line, header, and request body)",
+			Namespace: ns,
+			Buckets:   prometheus.LinearBuckets(10, 10, 10), // 10 buckets, each 10 bytes wide.
+		},
+		tags,
+	)
+
+	sc.bytesSent = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:      "bytes_sent",
+			Help:      "The the number of bytes sent to a client",
+			Namespace: ns,
+			Buckets:   prometheus.LinearBuckets(50, 25, 10), // 10 buckets, each 50 bytes wide.
```

Comment:
I would probably use exponential buckets for bytes sent, from 100 bytes to maybe 10 MiB.